### PR TITLE
Set `-DLIBOMP_FORTRAN_MODULES=ON` on windows

### PR DESCRIPTION
**Pull request by h-vetinari**
If we want to build OpenBLAS with OpenMP support with flang (c.f. https://github.com/conda-forge/openblas-feedstock/pull/115), we'll need to package the Fortran bits of OpenMP, otherwise we run into:
```
[...]
[3646/19184] Building Fortran object CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_2stage.f.obj
[3647/19184] Building Fortran object CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F.obj
FAILED: CMakeFiles/LAPACK_OVERRIDES.dir/lapack-netlib/SRC/ssytrd_sb2st.F.obj 
%BUILD_PREFIX%\Library\bin\flang.exe -I%SRC_DIR%\lapack-netlib\SRC -I%SRC_DIR%\lapack-netlib\LAPACKE\include -fopenmp -fopenmp -ffixed-line-length-72 -o CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F.obj -c CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F-pp.f
error: Semantic errors in CMakeFiles\LAPACK_OVERRIDES.dir\lapack-netlib\SRC\ssytrd_sb2st.F-pp.f
D:\\bld\\openblas_1719536014536\\work\\lapack-netlib\\SRC\\ssytrd_sb2st.F:237:11: error: Cannot read module file for module 'omp_lib': Source file 'omp_lib.mod' was not found
        use omp_lib
            ^^^^^^^
```

Flang 18 is going to be [unusable](https://github.com/llvm/llvm-project/issues/86463) to build OpenBLAS, but I'd like to open this for discussion/review already, so we can figure this out before LLVM 19 releases.

PS. This is rebased from the `dev` branch, with which I was able to get CI green in https://github.com/conda-forge/openblas-feedstock/pull/115.

### CODE REVIEW

**Comment by h-vetinari** [comment]
This is quite curious. We cannot build the fortran bits of OpenMP 18 with flang 18, but it works if _either_ of the versions is bumped to 19. In this PR we bump the flang version, on the `dev` branch it's the other way 'round.

In any case, even if we don't merge this for OpenMP 18, I want to enable this for OpenMP 19, so I'd appreciate feedback @conda-forge/openmp

**Comment by h-vetinari** [comment]
Gentle ping on this @conda-forge/openmp

**Comment by isuruf** [comment]
Can we move the `.mod` file to a different output? The cycles are also there in linux if we switch `_openmp_mutex`.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -21,26 +21,32 @@ source:
...
+      # due to https://github.com/conda/conda-build/issues/5403
+      overlinking_ignore_patterns:  # [ppc64le]
+        - "**"                      # [ppc64le]
       ignore_run_exports:
-        - libgcc-ng
-        - libstdcxx-ng
         - _openmp_mutex
+      ignore_run_exports_from:
+        - {{ compiler("cxx") }}
+        - {{ compiler("fortran") }}
```
Comment (isuruf) [inline recipe/meta.yaml:38]: The `.mod` files are specific to the gfortran version (maybe not the major version, but the soversion. not sure), so it's not possible to remove these.

**Comment by h-vetinari** [comment]
> Can we move the `.mod` file to a different output?

@isuruf, I've rebased this PR on #144 and added the output you wanted - I think I've incorporated all the feedback so far on this here. I still want to clean up the commit history better, but the _content_ of the changes should be OK now. As such, I'd ask you to please review this at your earliest convenience.

**Comment by h-vetinari** [comment]
Gentle ping @isuruf

File: recipe/meta.yaml [hunk 2]
```diff
@@ -71,6 +84,37 @@ outputs:
...
+        # for run-exports
+        - {{ stdlib("c") }}
+        - {{ compiler("cxx") }}
+        - {{ compiler("fortran") }}     # [linux]
+      host:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run_constrained:
+        - llvm-openmp {{ version }}
```
Comment (isuruf) [inline recipe/meta.yaml:110]:
```suggestion
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -71,6 +84,37 @@ outputs:
...
+        - cmake
+        # for run-exports
+        - {{ stdlib("c") }}
+        - {{ compiler("cxx") }}
+        - {{ compiler("fortran") }}     # [linux]
+      host:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run:
+        - {{ pin_subpackage("llvm-openmp", exact=True) }}
+      run_constrained:
```
Comment (isuruf) [inline recipe/meta.yaml:109]: We need a run_constrained on `flang` here as the mod file made here might only work with the flang compiler version used.

**Comment by h-vetinari** [comment]
Thanks for the review; should be fixed now.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -71,6 +85,37 @@ outputs:
...
+        - {{ compiler("fortran") }} =={{ fortran_compiler_version }}
```
Comment (isuruf) [inline recipe/meta.yaml:111]:
```suggestion
        - {{ compiler("fortran") }}
```

Comment (isuruf) [reply recipe/meta.yaml:111]: version is implicit in the jinja

Comment (h-vetinari) [reply recipe/meta.yaml:111]: Ah true, that already contains the version...

**Comment by h-vetinari** [comment]
Good now?

**Comment by h-vetinari** [comment]
I think we're close here @isuruf, can you please help me wrap this up?

**Pull request merged**